File: solutions/221.maximal-square.java
/*
 * @lc app=leetcode id=221 lang=java
 *
 * [221] Maximal Square
 */

// @lc code=start
class Solution {
    public int maximalSquare(char[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int[][] dp = new int[m][n];
        int maxSide = 0;
        for(int i=0; i<m; ++i) {
            dp[i][0] = matrix[i][0] - '0';
            maxSide = Math.max(maxSide, dp[i][0]);
        }
        for(int j=0; j<n; ++j) {
            dp[0][j] = matrix[0][j] - '0';
            maxSide = Math.max(maxSide, dp[0][j]);
        }
        for(int i=1; i<m; ++i) {
            for(int j=1; j<n; ++j) {
                if(matrix[i][j]=='1') {
                    dp[i][j] = Math.min(Math.min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1]) + 1;
                    maxSide = Math.max(maxSide, dp[i][j]); 
                }
            }
        }
        return maxSide*maxSide;
    }
}
// @lc code=end

File: solutions/221.maximal-square.rb
# @param {Character[][]} matrix
# @return {Integer}
def maximal_square(matrix)
  ans = 0
  m = matrix.size
  n = matrix.first.size

  (0...m).each do |x|
    (0...n).each do |y|
      if x > 0 && y > 0 && matrix[x][y].to_i == 1
        matrix[x][y] = 1 + min(matrix[x - 1][y], matrix[x][y - 1], matrix[x - 1][y - 1])
      end
      ans = max(ans, matrix[x][y])
    end
  end
  ans**2
end

def max(*args) = args.map(&:to_i).max
def min(*args) = args.map(&:to_i).min
